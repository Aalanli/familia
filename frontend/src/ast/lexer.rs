use std::{collections::VecDeque, fmt::Display, rc::Rc};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct Span {
    pub lhs: Loc,
    pub rhs: Loc,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct Loc {
    pub global: usize,
    pub line: u64,
    pub start: u32
}

impl Default for Loc {
    fn default() -> Self {
        Loc {
            global: 0,
            line: 0,
            start: 0,
        }
    }
}

impl Loc {
    pub fn span(&self, end: Loc) -> Span {
        Span {
            lhs: *self,
            rhs: end,
        }
    }
}

#[derive(Clone, PartialEq, Eq, Hash, Debug)]
pub struct Symbol {
    name: Rc<str>,
}

impl Symbol {
    pub fn from_str(s: &str) -> Self {
        Symbol { name: Rc::from(s) }
    }

    pub fn view(&self) -> &str {
        &*self.name
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Ident {
    pub name: Symbol,
    pub span: Span,
}

impl Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name.view())
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Tok {
    Class,
    Type,
    Fn,
    Let,
    Return,
    Plus,
    Equal,
    Dot,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Colon,
    SemiColon,
    ColonAcc, // ::
    Comma,
    I32,
    Ident(Ident),
    Int(i32),
}

#[derive(Debug)]
pub enum LexError {
    UnrecognizedToken(String),
    ParseIntError(String),
}

type Lex = Result<(Loc, Tok, Loc), LexError>;

pub struct Lexer<T: Iterator<Item = char>> {
    input: T,
    window: Option<char>,
    buf: VecDeque<char>,
    global_pos: usize,
    line_num: u64,
    line_pos: u32,
}

impl<T: Iterator<Item = char>> Lexer<T> {
    pub fn new(input: T) -> Self {
        Lexer {
            input,
            window: None,
            buf: VecDeque::new(),
            global_pos: 0,
            line_num: 0,
            line_pos: 0,
        }
    }

    fn next_char(&mut self) -> Option<char> {
        let c = self.input.next();
        if let Some(wc) = self.window {
            self.window = c;
            return Some(wc);
        }
        if let Some(c) = c {
            self.window = self.input.next();
            return Some(c);
        }
        None
    }
    
    fn peek_char(&mut self) -> Option<char> {
        if self.window.is_none() {
            self.window = self.input.next();
        }
        self.window
    }

    /// Fill the buffer while f return true
    fn fill_buf(&mut self, mut f: impl FnMut(char) -> bool) {
        while let Some(c) = self.peek_char() {
            if !f(c) {
                break;
            }
            self.buf.push_back(c);
            self.next_char();
        }
    }

    /// Clear the buffer while f return true
    fn clear_buf(&mut self, mut f: impl FnMut(char) -> bool) {
        while self.buf.len() > 0 {
            let c = self.buf[0];
            if !f(self.buf[0]) {
                break;
            }
            if c == '\n' {
                self.line_num += 1;
                self.line_pos = 0;
            } else {
                self.line_pos += 1;
            }
            self.global_pos += 1;
            self.buf.pop_front();
        }
    }

    fn clear_buf_by(&mut self, count: usize) {
        let mut i = count as isize;
        self.clear_buf(|_| {
            i -= 1;
            i >= 0
        });
    }

    fn match_buf(&self, s: &str) -> bool {
        if self.buf.len() != s.len() {
            return false;
        }
        for (i, c) in s.chars().enumerate() {
            if self.buf[i] != c {
                return false;
            }
        }
        true
    }

    fn match_and_eat(&mut self, s: &str) -> bool {
        if self.match_buf(s) {
            self.clear_buf_by(s.len());
            true
        } else {
            false
        }
    }

    fn fill_buf_by(&mut self, count: usize) {
        let mut i = 0;
        self.fill_buf(|_| {
            i += 1;
            i <= count
        });
    }

    fn fill_ident(&mut self) {
        self.fill_buf(|c| {
            c.is_alphanumeric()
        });
    }

    fn ignore_line(&mut self) {
        self.fill_buf(|c| c != '\n'); // stop when first '\n' is found
        self.fill_buf_by(1); // eat '\n'
        self.clear_buf(|_| true); // clear all characters
    }

    fn ignore_whitespace(&mut self) {
        self.clear_buf(|_| true);
        while let Some(c) = self.peek_char() {
            if c == ' ' {
                self.next_char();
                self.global_pos += 1;
                self.line_pos += 1;
            } else if c == '\n' {
                self.next_char();
                self.global_pos += 1;
                self.line_num += 1;
                self.line_pos = 0;
            } else {
                break;
            }
        }
    }

    fn is_ident(&self) -> bool {
        self.buf[0].is_alphabetic() && self.buf.iter().all(|c| c.is_alphanumeric())
    }

    fn get_loc(&self) -> Loc {
        Loc {
            global: self.global_pos,
            line: self.line_num,
            start: self.line_pos,
        }
    }

    fn next(&mut self) -> Option<Lex> {
        loop {
            self.ignore_whitespace();
            if self.peek_char().is_none() {
                return None;
            }
            self.fill_buf_by(1);
            let lpos = self.get_loc();
            if self.match_and_eat("{") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::LBrace, rpos)));
            } else if self.match_and_eat("}") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::RBrace, rpos)));
            } else if self.match_and_eat("(") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::LParen, rpos)));
            } else if self.match_and_eat(")") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::RParen, rpos)));
            } else if self.match_and_eat(":") {
                if self.peek_char() == Some(':') {
                    self.fill_buf_by(1);
                    self.clear_buf_by(1);
                    let rpos = self.get_loc();
                    return Some(Ok((lpos, Tok::ColonAcc, rpos)));
                }
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Colon, rpos)));
            } else if self.match_and_eat("+") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Plus, rpos)));
            } else if self.match_and_eat(";") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::SemiColon, rpos)));
            } else if self.match_and_eat(",") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Comma, rpos)));
            } else if self.match_and_eat(".") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Dot, rpos)));
            } else if self.match_and_eat("=") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Equal, rpos)));
            }

            if self.buf[0] == '/' && self.peek_char() == Some('/') {
                self.ignore_line();
                continue;
            }
            
            self.fill_ident();
            // println!("{:?}", self.buf.iter().collect::<String>());
            if self.match_and_eat("i32") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::I32, rpos)));
            } else if self.match_and_eat("class") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Class, rpos)));
            } else if self.match_and_eat("type") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Type, rpos)));
            } else if self.match_and_eat("fn") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Fn, rpos)));
            } else if self.match_and_eat("let") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Let, rpos)));
            } else if self.match_and_eat("return") {
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Return, rpos)));
            } else if self.buf.iter().all(|x| x.is_numeric()) { // todo: should change to regex
                let num = self.buf.iter().collect::<String>().parse().map_err(|x| LexError::ParseIntError(format!("{x}")));
                self.clear_buf(|_| true);
                if let Err(e) = num {
                    return Some(Err(e));
                }
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Int(num.unwrap()), rpos)));
            } else if self.is_ident() {
                let ident = self.buf.iter().collect::<String>();
                self.clear_buf(|_| true);
                let rpos = self.get_loc();
                return Some(Ok((lpos, Tok::Ident(Ident {
                    name: Symbol::from_str(&ident),
                    span: lpos.span(rpos),
                }), self.get_loc())));
            }

            let res = Some(Err(LexError::UnrecognizedToken(self.buf.iter().collect())));
            self.clear_buf(|_| true);
            return res;
        }
    }
}

impl<T: Iterator<Item = char>> Iterator for Lexer<T> {
    type Item = Lex;

    fn next(&mut self) -> Option<Self::Item> {
        self.next()
    }
}


#[cfg(test)]
mod test_lex {
    use super::*;
    #[test]
    fn test_lex1() {
        let input = "fn main() { return 1 + 2; }";

        let lexer = Lexer::new(input.chars());
        for i in lexer {
            i.unwrap();
        }
    }

    #[test]
    fn test_lex2() {
        let input = "\
            type T = {
                a: i32,
                b: i32
            }

            class c1 {
                fn foo(a: T, b: i32): i32 {
                    return a.a + b;
                }
            }

            fn main() {
                c1::foo({a: 1, b: 2}, 3);
            }
        ";

        let lexer = Lexer::new(input.chars());
        for i in lexer {
            i.unwrap();
        }
    }

    #[test]
    fn test_lex3() {
        let input1 = "f124";
        let input2 = "124";
        let input3 = "fn";
        let input4 = "fn1";
        let input5 = "a::b::c3::2";

        let mut lexer1 = Lexer::new(input1.chars());
        if let Tok::Ident(..) = lexer1.next().unwrap().unwrap().1 {
        } else {
            panic!();
        }
        
        let mut lexer2 = Lexer::new(input2.chars());
        assert!(lexer2.next().unwrap().unwrap().1 == Tok::Int(124));
        let mut lexer3 = Lexer::new(input3.chars());
        assert!(lexer3.next().unwrap().unwrap().1 == Tok::Fn);
        let mut lexer4 = Lexer::new(input4.chars());
        if let Tok::Ident(..) = lexer4.next().unwrap().unwrap().1 {
        } else {
            panic!();
        }

        let mut lexer5 = Lexer::new(input5.chars());
        if let Tok::Ident(..) = lexer5.next().unwrap().unwrap().1 {
        } else {
            panic!();
        }
        assert!(lexer5.next().unwrap().unwrap().1 == Tok::ColonAcc);
    }
}

