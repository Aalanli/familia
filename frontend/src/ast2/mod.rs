use std::ops::ControlFlow;

use crate::ProgramSource;
use derive_new::new;

pub use super::lexer::Span;

#[allow(non_snake_case)]
pub fn P<T>(x: T) -> Box<T> {
    Box::new(x)
}

#[salsa::interned]
pub struct Symbol<'db> {
    #[return_ref]
    pub sym: String,
}

#[salsa::interned]
pub struct Path<'db> {
    #[return_ref]
    pub symbols: Vec<Symbol<'db>>,
}

#[derive(Eq, PartialEq, Copy, Clone, Hash, Debug, salsa::Update)]
pub struct PathSpan<'db> {
    pub path: Path<'db>,
    pub span: Span,
}

#[derive(Eq, PartialEq, Copy, Clone, Hash, Debug, salsa::Update)]
pub struct SymbolSpan<'db> {
    pub sym: Symbol<'db>,
    pub span: Span,
}

#[salsa::interned]
pub struct Type<'db> {
    #[return_ref]
    pub kind: TypeKind<'db>,
    pub span: Span,
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub enum TypeKind<'db> {
    Void,
    Struct { fields: Vec<TypedVar<'db>> },
    Symbol(Path<'db>),
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub struct Var<'db> {
    pub name: SymbolSpan<'db>,
    pub ty: Option<Type<'db>>,
    pub span: Span,
}

impl<'db> Var<'db> {
    pub fn into_tvar(self, db: &dyn crate::Db) -> Option<TypedVar<'db>> {
        if self.ty.is_none() {
            crate::Diagnostic::report_syntax_err(db, "expected type annotation");
        }
        Some(TypedVar {
            name: self.name,
            ty: self.ty?,
            span: self.span,
        })
    }
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub struct TypedVar<'db> {
    pub name: SymbolSpan<'db>,
    pub ty: Type<'db>,
    pub span: Span,
}

#[salsa::tracked]
pub struct Module<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    #[return_ref]
    pub body: Vec<Decls<'db>>,

    pub src: ProgramSource,
}

#[salsa::tracked]
pub struct InterfaceDecl<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    #[return_ref]
    pub body: Vec<Decls<'db>>,
}

#[salsa::tracked]
pub struct ClassImpl<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    pub for_it: Option<Path<'db>>,
    #[return_ref]
    pub body: Vec<Decls<'db>>,
}

#[salsa::tracked]
pub struct FnDecl<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    #[return_ref]
    pub args: Vec<Var<'db>>,
    pub res_ty: Type<'db>,
}

#[salsa::tracked]
pub struct FnImpl<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    #[return_ref]
    pub args: Vec<Var<'db>>,
    pub res_ty: Type<'db>,
    #[return_ref]
    pub body: Vec<Stmt<'db>>,
}

#[salsa::tracked]
pub struct TypeDecl<'db> {
    #[id]
    pub name: Symbol<'db>,
    pub name_span: Span,
    pub type_: Type<'db>,
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub enum Decls<'db> {
    Interface(InterfaceDecl<'db>),
    Class(ClassImpl<'db>),
    FnDecl(FnDecl<'db>),
    FnImpl(FnImpl<'db>),
    TyDecl(TypeDecl<'db>),
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub struct Stmt<'db> {
    pub kind: StmtKind<'db>,
    pub span: Span,
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub enum StmtKind<'db> {
    LetStmt { var: Var<'db>, expr: Expr<'db> },
    ExprStmt(Expr<'db>),
    AssignStmt { lhs: Expr<'db>, rhs: Expr<'db> },
    ReturnStmt(Expr<'db>),
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub struct Expr<'db> {
    pub kind: ExprKind<'db>,
    pub span: Span,
}

#[derive(Eq, PartialEq, Clone, Hash, Debug, salsa::Update)]
pub enum ExprKind<'db> {
    Var(Var<'db>),
    VoidLit,
    IntLit(i32),
    StringLit(Symbol<'db>),
    MethodCall {
        expr: Box<Expr<'db>>,
        sym: SymbolSpan<'db>,
        args: Vec<Expr<'db>>,
    },
    GetAttr {
        expr: Box<Expr<'db>>,
        sym: SymbolSpan<'db>,
    },
    Add {
        lhs: Box<Expr<'db>>,
        rhs: Box<Expr<'db>>,
    },
    Call {
        path: PathSpan<'db>,
        args: Vec<Expr<'db>>,
    },
    Struct {
        args: Vec<(SymbolSpan<'db>, Expr<'db>)>,
    },
}

pub trait Visitor<'db>: Sized {
    type Result: VisitorResult;

    fn visit_stmt(&mut self, stmt: &Stmt<'db>) -> Self::Result {
        walk_stmt(self, stmt)
    }
}

pub fn walk_stmt<'db, V: Visitor<'db>>(visitor: &mut V, stmt: &Stmt<'db>) -> V::Result {
    todo!()
}

/// Similar to the `Try` trait, but also implemented for `()`.
pub trait VisitorResult {
    type Residual;
    fn output() -> Self;
    fn from_residual(residual: Self::Residual) -> Self;
    fn from_branch(b: ControlFlow<Self::Residual>) -> Self;
    fn branch(self) -> ControlFlow<Self::Residual>;
}

impl VisitorResult for () {
    #[cfg(feature = "nightly")]
    type Residual = !;

    #[cfg(not(feature = "nightly"))]
    type Residual = core::convert::Infallible;

    fn output() -> Self {}
    fn from_residual(_: Self::Residual) -> Self {}
    fn from_branch(_: ControlFlow<Self::Residual>) -> Self {}
    fn branch(self) -> ControlFlow<Self::Residual> {
        ControlFlow::Continue(())
    }
}

impl<T> VisitorResult for ControlFlow<T> {
    type Residual = T;

    fn output() -> Self {
        ControlFlow::Continue(())
    }
    fn from_residual(residual: Self::Residual) -> Self {
        ControlFlow::Break(residual)
    }
    fn from_branch(b: Self) -> Self {
        b
    }
    fn branch(self) -> Self {
        self
    }
}

#[macro_export]
macro_rules! try_visit {
    ($e:expr) => {
        match $crate::visit::VisitorResult::branch($e) {
            core::ops::ControlFlow::Continue(()) => (),
            #[allow(unreachable_code)]
            core::ops::ControlFlow::Break(r) => {
                return $crate::visit::VisitorResult::from_residual(r);
            }
        }
    };
}

#[macro_export]
macro_rules! visit_opt {
    ($visitor: expr, $method: ident, $opt: expr $(, $($extra_args: expr),* )?) => {
        if let Some(x) = $opt {
            $crate::try_visit!($visitor.$method(x $(, $($extra_args,)* )?));
        }
    }
}

#[macro_export]
macro_rules! walk_list {
    ($visitor: expr, $method: ident, $list: expr $(, $($extra_args: expr),* )?) => {
        for elem in $list {
            $crate::try_visit!($visitor.$method(elem $(, $($extra_args,)* )?));
        }
    }
}

#[macro_export]
macro_rules! walk_visitable_list {
    ($visitor: expr, $list: expr $(, $($extra_args: expr),* )?) => {
        for elem in $list {
            $crate::try_visit!(elem.visit_with($visitor $(, $($extra_args,)* )?));
        }
    }
}

#[cfg(test)]
mod test_ast {
    #[salsa::interned]
    struct Foo<'db> {
        pub name: String,
        #[no_eq]
        pub idx: i32
    }

    #[test]
    fn test_intern_eq() {
        let db = salsa::default_database();
        let foo1 = Foo::new(&db, "name".into(), 1);
        let foo2 = Foo::new(&db, "name".into(), 2);
        let foo3 = Foo::new(&db, "name".into(), 1);
        assert!(foo1 != foo2);
        assert!(foo1 == foo3);
    }
}