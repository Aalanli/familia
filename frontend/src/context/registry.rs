use std::cell::RefCell;
use std::collections::HashMap;
use std::hash::Hash;

use anyhow::{anyhow, Result};

#[derive(Debug)]
pub struct Registry<Idx, T: ?Sized> {
    data: RefCell<HashMap<Idx, Box<T>>>,
}

impl<Idx: Eq + Hash + Copy, T: ?Sized> Registry<Idx, T> {
    pub fn new() -> Self {
        Registry {
            data: RefCell::new(HashMap::new()),
        }
    }

    pub fn get(&self, id: Idx) -> Option<&T> {
        let data = &*self.data.borrow();
        if let Some(ptr) = data.get(&id).map(|x| x) {
            // this is safe because we return a reference to the heap, whose lifetime is the same as the registry
            Some(unsafe { &*(ptr.as_ref() as *const T) })
        } else {
            None
        }
    }

    pub fn get_mut(&mut self, id: Idx) -> Option<&mut T> {
        let data = &mut *self.data.borrow_mut();
        if let Some(ptr) = data.get_mut(&id).map(|x| x) {
            Some(unsafe { &mut *(ptr.as_mut() as *mut T) })
        } else {
            None
        }
    }

    pub fn insert_with(&self, id: Idx, value: Box<T>) -> Result<()> {
        let mut data = self.data.borrow_mut();
        if data.contains_key(&id) {
            return Err(anyhow!("cannot insert duplicate key"));
        }

        data.insert(id, value);
        Ok(())
    }

    pub fn get_insert_with(&self, id: Idx, f: impl FnOnce() -> Box<T>) -> &T {
        if let Some(x) = self.get(id) {
            return x;
        } else {
            let mut data = self.data.borrow_mut();
            if !data.contains_key(&id) {
                data.insert(id, f());
            }
        }
        self.get(id).unwrap()
    }

    pub fn pop(&mut self, id: Idx) -> Option<Box<T>> {
        self.data.borrow_mut().remove(&id)
    }
}
