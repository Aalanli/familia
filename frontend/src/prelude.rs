pub use crate::ast::Span;
pub use crate::error as err;
pub use crate::{ModSource, PhaseResult};
pub use err::{ErrorType, PResult, ProgramError};
