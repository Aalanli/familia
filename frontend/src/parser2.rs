use crate::ast2::*;
use crate::context::Db;
use crate::lexer::Lexer;
use crate::ProgramSource;

use ariadne::{sources, Color, Label, Report, ReportKind};
use chumsky::prelude::*;
use either::Either;
use std::{collections::HashMap, env, fmt, fs};

pub type Span = SimpleSpan<usize>;

#[derive(Clone, Debug, PartialEq)]
enum Token<'src> {
    I32(i32),
    Str(&'src str),
    Ctrl(char),
    Ident(&'src str),
    Op(&'src str),
    Fn,
    Let,
    Return,
    Class,
    Interface,
}

impl<'src> fmt::Display for Token<'src> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Token::I32(i) => write!(f, "{}", i),
            Token::Str(s) => write!(f, "{}", s),
            Token::Op(s) => write!(f, "{}", s),
            Token::Ctrl(c) => write!(f, "{}", c),
            Token::Ident(s) => write!(f, "{}", s),
            Token::Fn => write!(f, "fn"),
            Token::Let => write!(f, "let"),
            Token::Return => write!(f, "return"),
            Token::Class => write!(f, "class"),
            Token::Interface => write!(f, "interface"),
        }
    }
}

fn lexer<'src>(
) -> impl Parser<'src, &'src str, Vec<(Token<'src>, Span)>, extra::Err<Rich<'src, char, Span>>> {
    let num = text::int(10)
        .to_slice()
        .from_str()
        .unwrapped()
        .map(Token::I32);

    let str_ = just('"')
        .ignore_then(none_of('"').repeated())
        .then_ignore(just('"'))
        .to_slice()
        .map(Token::Str);

    // A parser for operators
    let op = one_of("+*-/!=")
        .repeated()
        .at_least(1)
        .to_slice()
        .map(Token::Op);

    let ctrl = one_of("()[]{}:;,.").map(Token::Ctrl);

    let ident = text::ascii::ident().map(|ident: &str| match ident {
        "fn" => Token::Fn,
        "let" => Token::Let,
        "return" => Token::Return,
        "class" => Token::Class,
        "interface" => Token::Interface,
        _ => Token::Ident(ident),
    });

    let token = num.or(str_).or(op).or(ctrl).or(ident);

    let comment = just("//")
        .then(any().and_is(just('\n').not()).repeated())
        .padded();

    token
        .map_with(|tok, e| (tok, e.span()))
        .padded_by(comment.repeated())
        .padded()
        // If we encounter an error, skip and attempt to lex the next character as a token instead
        .recover_with(skip_then_retry_until(any().ignored(), end()))
        .repeated()
        .collect()
}

type ParserInput<'tokens, 'src> =
    chumsky::input::SpannedInput<Token<'src>, Span, &'tokens [(Token<'src>, Span)]>;

fn symbol_span<'db, 'tokens: 'db, 'src: 'tokens>(
    db: &'db Db,
) -> impl Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    SymbolSpan,
    extra::Err<Rich<'tokens, Token<'src>, Span>>,
> + Clone
       + 'db {
    let ident = select! { Token::Ident(i) => i };
    ident.map_with(move |ident, e| {
        let sym = Symbol::new(db, ident.to_string());
        WithSpan::new(sym, e.span())
    })
}

fn parse_path<'db, 'tokens: 'db, 'src: 'tokens>(
    db: &'db Db,
) -> impl Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    Path,
    extra::Err<Rich<'tokens, Token<'src>, Span>>,
> + Clone
       + 'db {
    let ident = select! { Token::Ident(i) => i };
    let raw_symbol = ident.map(|ident| {
        let sym = Symbol::new(db, ident.to_string());
        sym
    });

    let path = raw_symbol
        .separated_by(just(Token::Ctrl('.')))
        .at_least(1)
        .collect::<Vec<_>>()
        .map(|x| Path::new(db, x));
    path
}

fn type_parser<'db, 'tokens: 'db, 'src: 'tokens>(
    db: &'db Db,
) -> impl Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    TypeId,
    extra::Err<Rich<'tokens, Token<'src>, Span>>,
> + Clone
       + 'db {
    recursive(|type_parser| {
        let void = just(Token::Ctrl('('))
            .ignore_then(just(Token::Ctrl(')')))
            .map(|_| TypeKind::Void);

        let typed_var = symbol_span(db)
            .then_ignore(just(Token::Ctrl(':')))
            .then(type_parser.clone())
            .map(|(s, t)| TypedVar::new(s, t));

        let structure = typed_var
            .separated_by(just(Token::Ctrl(',')))
            .allow_trailing()
            .at_least(1)
            .collect::<Vec<_>>()
            .map(|x| TypeKind::Struct { fields: x })
            .delimited_by(just(Token::Ctrl('{')), just(Token::Ctrl('}')))
            .labelled("struct");

        void.or(parse_path(db).map(|p| TypeKind::Symbol(p)))
            .or(structure)
            .map_with(|tykind, e| TypeId::new(db, WithSpan::new(tykind, e.span())))
    })
}

fn expr_parser<'db, 'tokens: 'db, 'src: 'tokens>(
    db: &'db Db,
) -> impl Parser<
    'tokens,
    ParserInput<'tokens, 'src>,
    Expr,
    extra::Err<Rich<'tokens, Token<'src>, Span>>,
> + Clone
       + 'db {
    recursive(|expr| {
        let ty_hint = just(Token::Ctrl(':')).ignore_then(type_parser(db)).or_not();

        let var = symbol_span(db)
            .then(ty_hint)
            .map(|(s, ty)| ExprKind::Var(Var { name: s, ty }));

        let void_lit = just(Token::Ctrl('('))
            .ignored()
            .then_ignore(just(Token::Ctrl(')')))
            .map(|_| ExprKind::VoidLit);

        let int_lit = select! {Token::I32(i) => i}.map(|i| ExprKind::IntLit(i));
        let str_lit = select! { Token::Str(s) => s }
            .map(|s| ExprKind::StringLit(Symbol::new(db, s.to_string())));

        let items = expr
            .clone()
            .separated_by(just(Token::Ctrl(',')))
            .allow_trailing()
            .collect::<Vec<_>>();

        let atom = var
            .or(void_lit)
            .or(int_lit)
            .or(str_lit)
            .map_with(|kind, e1| Expr {
                kind,
                span: e1.span(),
            })
            .or(expr
                .clone()
                .delimited_by(just(Token::Ctrl('(')), just(Token::Ctrl(')'))));

        let call = atom.foldl_with(
            items
                .delimited_by(just(Token::Ctrl('(')), just(Token::Ctrl(')')))
                .map(Either::Left)
                .or(just(Token::Ctrl('.'))
                    .ignore_then(symbol_span(db))
                    .map(Either::Right))
                .repeated(),
            |f, a, e| {
                let kind = match a {
                    Either::Left(args) => ExprKind::Call {
                        object: Box::new(f),
                        args,
                    },
                    Either::Right(attr) => ExprKind::GetAttr {
                        expr: Box::new(f),
                        sym: attr,
                    },
                };
                Expr {
                    kind,
                    span: e.span(),
                }
            },
        );

        let op = just(Token::Op("+")).ignored();
        let add = call
            .clone()
            .foldl_with(op.then(call).repeated(), |a, (_, b), e| Expr {
                kind: ExprKind::Add {
                    lhs: Box::new(a),
                    rhs: Box::new(b),
                },
                span: e.span(),
            });

        let expr = add;

        expr
    })
}

#[cfg(test)]
mod test_parse {
    use fmt::Display;
    use insta::assert_debug_snapshot;

    use crate::Diagnostic;

    use super::*;

    macro_rules! run_parser {
        ($src:literal, $parser:expr) => {{
            let mut w = Vec::new();
            let (tokens, err) = lexer().parse(&($src)).into_output_errors();
            let errs = err
                .into_iter()
                .map(|e| e.map_token(|c| c.to_string()))
                .collect::<Vec<_>>();
            if tokens.is_none() {
                display_err(&mut w, $src.to_string(), "L".to_string(), &errs);
            } else {
                let (item, err_parse) = ($parser)
                    .parse(
                        tokens
                            .as_ref()
                            .unwrap()
                            .as_slice()
                            .spanned(($src.len()..$src.len()).into()),
                    )
                    .into_output_errors();
                if let Some(item) = item {
                    use std::io::Write;
                    write!(w, "{:?}", item).unwrap();
                } else {
                    display_err(&mut w, $src.to_string(), "L".to_string(), &err_parse);
                }
            }

            String::from_utf8(w).unwrap()
        }};
    }

    #[test]
    fn test_lex1() {
        let (tokens, errs) = lexer()
            .parse("input a b ++ \"abc\" -+= let a return : { [")
            .into_output_errors();
        assert_debug_snapshot!((tokens, errs));
    }

    fn display_err<T: Display>(
        mut w: impl std::io::Write,
        src: String,
        filename: String,
        err: &[Rich<T>],
    ) {
        // .map(|e| e.map_token(|tok| tok.to_string()))
        for e in err.into_iter() {
            Report::build(ReportKind::Error, filename.clone(), e.span().start)
                .with_message(e.to_string())
                .with_label(
                    Label::new((filename.clone(), e.span().into_range()))
                        .with_message(e.reason().to_string()), // .with_color(Color::Red),
                )
                .with_labels(e.contexts().map(|(label, span)| {
                    Label::new((filename.clone(), span.into_range()))
                        .with_message(format!("while parsing this {}", label))
                    // .with_color(Color::Yellow)
                }))
                .finish()
                .write(sources([(filename.clone(), src.clone())]), &mut w)
                .unwrap();
        }
    }

    #[test]
    fn test_parse_type() {
        let db = Db::new();
        let res = run_parser!("{a: i32}", type_parser(&db));
        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_parse_expr() {
        let db = Db::new();
        let res = run_parser!("((a.a(2, \"3\") + 3)()).c", expr_parser(&db));
        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_parse_getattr() {
        let db = Db::new();
        let res = run_parser!("a.b.c.d.e", expr_parser(&db));
        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_parse_getattr_fail() {
        let db = Db::new();
        let res = run_parser!("a().b", expr_parser(&db));
        assert_debug_snapshot!(res);
    }
}
