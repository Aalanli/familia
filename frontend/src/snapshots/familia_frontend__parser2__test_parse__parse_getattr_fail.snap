---
source: frontend/src/parser2.rs
expression: res
---
"Expr { kind: GetAttr { expr: Expr { kind: Call { object: Expr { kind: Var(Var { name: WithSpan { x: Symbol(NodeId(0)), span: 0..1 }, ty: None }), span: 0..1 }, args: [] }, span: 0..3 }, sym: WithSpan { x: Symbol(NodeId(1)), span: 4..5 } }, span: 0..5 }"
